python/tvm/auto_scheduler/relay_integration.py

```diff
@@ -0,0 +1,179 @@
+        The module or function to tune
+    params: dict of str to numpy array
+        The associated parameters of the program
+    target: tvm.target.Target
+        The compilation target
+
+    Returns
+    -------
+    wkl_keys: List[str]
+        The hash key of extracted workloads
+    wkl_weights: List[int]
+        The weight (i.e. the number of appearance) of extracted workload
+    """
+    # pylint: disable=import-outside-toplevel
+    from tvm import relay
+
+    env = TracingEnvironment()
+    with env:
+        # Run the compiler to collect all TOPI calls during compilation.
+
+        # Wrap build call in a new thread to avoid the conflict
+        # between python's multiprocessing and tvm's thread pool
+        build_thread = threading.Thread(target=call_all_topi_funcs, args=(mod, params, target))
+        build_thread.start()
+        build_thread.join()
+        relay.backend.compile_engine.get().clear()
+
+    # create tasks for target
+    wkl_keys = []
+    wkl_weights = []
+    for wkl_key, wkl_weight in env.wkl_key_collection.items():
+        wkl_keys.append(wkl_key)
+        wkl_weights.append(wkl_weight)
+
+    return wkl_keys, wkl_weights
+
+
+class TracingEnvironment:
+    """Global environment for tracing all topi function calls"""
+
+    current = None
+
+    def __init__(self):
+        self.relay_disable_build_cache = "false"
+        self.layout_rewrite_success_ct = 0
+        self.wkl_key_collection = {}
+
+    def __enter__(self):
+        self.relay_disable_build_cache = os.environ.get("TVM_RELAY_DISABLE_BUILD_CACHE", "false")
+        os.environ["TVM_RELAY_DISABLE_BUILD_CACHE"] = "true"
+        TracingEnvironment.current = self
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        os.environ["TVM_RELAY_DISABLE_BUILD_CACHE"] = self.relay_disable_build_cache
+        TracingEnvironment.current = None
+
+    def add_workload_key(self, key):
+        """Add the workload key of an Ansor search task
+
+        Parameters
+        ----------
+        key: str
+            The workload key of a task
+        """
+        if key in self.wkl_key_collection:
+            self.wkl_key_collection[key] += 1
+        else:
+            self.wkl_key_collection[key] = 1
+
+
+def traverse_to_get_io_tensors(outs):
+    """Traverse from a list of output tensors to get a whole computational DAG"""
+    inputs = []
+
+    visited = set()
+
+    def traverse(t):
+        if t in visited:
+            return
+        if isinstance(t.op, PlaceholderOp):
+            inputs.append(t)
+        elif isinstance(t.op, ComputeOp):
+            for x in t.op.input_tensors:
+                traverse(x)
+        visited.add(t)
+
+    for t in outs:
+        traverse(t)
+
+    return inputs + list(outs)
+
+
+# The suffix of implementations that use the auto-scheduler in the OpStrategy.
+auto_schedule_impl_suffix = ".auto_scheduler"
+
+
+def auto_schedule_topi(outs):
```

Comment:
Is it better to call it `auto_schedule_te` (and replace all `topi` in this file with `te`)? I imagine there are cases where inputs to this function don't come from topi, such as TE from gradient ops generated by Relay.

Reply:
We cannot call it `auto_schedule_te`. We have another function `auto_schedule` (https://github.com/apache/incubator-tvm/blob/e59c603515befb02035e237794aa0645dbfbaf09/python/tvm/auto_scheduler/auto_schedule.py#L161)  for your use case.
But this function is designed to be used as a TOPI schedule function for Relay.